bin/quality_controls.qmd

```diff
@@ -8,57 +8,88 @@ jupyter: python3
 
 # Introduction
 
+<!-- TODO: Add SpatialData in intro -->
 Spatial Transcriptomics data analysis involves several steps, including quality
 controls (QC) and pre-processing, to ensure the reliability of downstream
-analyses. This is an essential step in spatial transcriptomics to
-identify and filter out spots and genes that may introduce noise and/or bias
-into the analysis.
+analyses. This is an essential step in spatial transcriptomics to identify and
+filter out spots and genes that may introduce noise and/or bias into the
+analysis.
 
 This report outlines the QC and pre-processing steps for Visium Spatial
 Transcriptomics data using the [AnnData format](https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html)
 and the [`scanpy` Python package](https://scanpy.readthedocs.io/en/stable/).
 The anndata format is utilized to organize and store the Spatial Transcriptomics
 data. It includes information about counts, features, observations, and
 additional metadata. The anndata format ensures compatibility with various
-analysis tools and facilitates seamless integration into existing workflows.
+analysis tools and facilitates seamless integration into existing workflows. The AnnData object is saved in a zarr [SpatialData object](https://spatialdata.scverse.org/en/latest/design_doc.html#table-table-of-annotations-for-regions).
 
 ```{python}
 #| tags: [parameters]
 #| echo: false
-input_adata_raw = "st_adata_raw.h5ad"  # Name of the input anndata file
+input_sdata = "sdata_raw.zarr"  # Input: SpatialData file
 min_counts = 500  # Min counts per spot
 min_genes = 250  # Min genes per spot
 min_spots = 1  # Min spots per gene
 mito_threshold = 20  # Mitochondrial content threshold (%)
 ribo_threshold = 0  # Ribosomal content threshold (%)
 hb_threshold = 100  # content threshold (%)
-output_adata_filtered = "st_adata_filtered.h5ad"  # Name of the output anndata file
+artifact_dir = "artifacts"
+output_adata = "adata_filtered.h5ad"  # Output: AnnData file
+output_sdata = "sdata_filtered.zarr"  # Output: SpatialData file
 ```
 
 ```{python}
-import scanpy as sc
-import scipy
 import pandas as pd
 import matplotlib.pyplot as plt
+import numpy as np
+import os
+import scanpy as sc
+import scipy
 import seaborn as sns
+import spatialdata
+from anndata import AnnData
 from IPython.display import display, Markdown
 from textwrap import dedent
 plt.rcParams["figure.figsize"] = (6, 6)
 ```
 
+```{python}
+# Make sure we can use scanpy plots with the AnnData object exported from sdata.table
+# This code is taken from the early version of https://github.com/scverse/spatialdata-io/pull/102/
```

Comment:
It seems that this PR is now merged and the latest release is `0.1.2` (where are currently using `0.0.9`), so maybe this can be changed now?

Reply:
It would definitely be better, I agree.
I tried to use the new `spatialdata.experimental.to_legacy_anndata()` but without any success for now.
It seems to have trouble saving images with "downscaled_lowres" coordinate system, asking me to provide a `coordinate_system` even though it is provided:
```python
>>> spatialdata = spatialdata_io.visium(
        args.SRCountDir, counts_file="raw_feature_bc_matrix.h5", dataset_id="visium"
    )
>>> adata = to_legacy_anndata(sdata, include_images=True, coordinate_system="downscaled_hires")
/opt/conda/lib/python3.10/site-packages/spatialdata/_core/query/relational_query.py:325: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  groups_df = table.obs.groupby(by=region_column_name)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/conda/lib/python3.10/site-packages/spatialdata_io/converters/legacy_anndata.py", line 176, in to_legacy_anndata
    bb = get_extent(sdata_images)
  File "/opt/conda/lib/python3.10/functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/opt/conda/lib/python3.10/site-packages/spatialdata/_core/data_extent.py", line 239, in _
    raise ValueError(
ValueError: The SpatialData object does not contain any element in the coordinate system 'global', please pass a different coordinate system with the argument 'coordinate_system'.
```
Maybe I should open an issue in spatialdata-io.

Reply:
Okay, then we can keep the custom code for now and change it when there is a solution to the issue.